### ISSUE TITLE
pax-logging.jar is creating problems for apps that use slf4j

### ISSUE THREAD

**Issue by phillipross** [payara/Payara#652]
It appears that pax-logging dependency was introduced between v4.1.1.154 and v4.1.1.161.
This is creating issues with apps that use SLF4J since pax-logging binds pax logging framework to SLF4J before apps have a chance to register their own bindings.

In my testing... web modules can include a glassfish-web.xml element to disable classloader delegation:
```
<!DOCTYPE glassfish-web-app PUBLIC "-//GlassFish.org//DTD GlassFish Application Server 3.1 Servlet 3.0//EN" "http://glassfish.org/dtds/glassfish-web-app_3_0-1.dtd">
<glassfish-web-app>
    <class-loader delegate="false"/>
</glassfish-web-app>
```
But this will not work for modules that are not inside the web modules.... such as ear modules and other libraries in the lib folder of an EAR file.  I'm not aware of a way to fix this for these later cases.  I have many apps which package slf4j and logback in th EAR's lib directory, but with this pax-logging now in 4.1.1.161, I get deployment errors that show a vey confused SLF4J:
```
Caused by: java.lang.ClassCastException: org.ops4j.pax.logging.slf4j.Slf4jLoggerFactory cannot be cast to ch.qos.logback.classic.LoggerContext
```

I *AM* able to work around this so far by hacking the pax-logging.jar module in the 4.1.1.161 modules... removing the `org/ops4j/pax/logging/slf4j` directory and its contents from the jar, but I don't know what side-effects that's going to have :fearful:

**Comment by phillipross** [comment]
Ultimately I think #419 is what we need to work around problems like this.

**Comment by phillipross** [comment]
As an experiment, I cloned master, commented out all dependency declarations across the pom files for `pax-logging`, then did a full build (WITHOUT -DskipTests=true) and the build worked flawlessly.  After installing this build... slf4j+logback works fine from lib directry in ear files.

**Comment by dorpsidioot** [comment]
Having the same issue with apache wicket's logging, also uses SLF4J and my server log now gets polluted with:
[#|2016-02-08T08:58:01.444+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281444;_LevelValue=800;|org.apache.wicket.core.util.resource.locator.ResourceStreamLocator|#]

[#|2016-02-08T08:58:01.444+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281444;_LevelValue=800;|] :|#]

[#|2016-02-08T08:58:01.444+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281444;_LevelValue=800;|Attempting to locate resource 'org/apache/wicket/markup/html/form/wicket-package_nl_BE.properties.xml' using finder'[classpath: META-INF/resources/]'|#]

[#|2016-02-08T08:58:01.448+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281448;_LevelValue=800;|[|#]

[#|2016-02-08T08:58:01.448+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281448;_LevelValue=800;|org.apache.wicket.core.util.resource.locator.ResourceStreamLocator|#]

[#|2016-02-08T08:58:01.448+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281448;_LevelValue=800;|] :|#]

[#|2016-02-08T08:58:01.459+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281459;_LevelValue=800;|Attempting to locate resource 'org/apache/wicket/markup/html/form/wicket-package_nl.properties' using finder'[classpath]'|#]

[#|2016-02-08T08:58:01.462+0100|INFO|Payara 4.1||_ThreadID=186;_ThreadName=http-listener-1(1);_TimeMillis=1454918281462;_LevelValue=800;|[|#]

**Comment by dorpsidioot** [comment]
Current workaround I found is to add a JVM setting do put the pax-logging default level to INFO to avoid the spammy nature of this stuff. This clears out all the superfluous debug logging, but the main issue is still there.

**Comment by ke5stl** [comment]
what is lost by removing the pax-logging.jar from the glassfish/modules entirely?  what I found was that the pax-logging seems to be something to do with some testing capability for embedded uses.  seems like a fairly narrow need.  I tried removing pax-logging.jar entirely, and have not seen a problem so far....

my particular issue is that the pax-logging jar has within it a version of log4j classes or interfaces which don't match what my application is using, and this totally kills my app (since the pax-logging classes are missing some constructor variants).  removing pax-logging at least lets the app deploy and start.

**Comment by phillipross** [comment]
I believe pax-logging was introduced as a hasty way of addressing a missing dependency of JCL "commons logging" that newer versions of metro were looking for.  I'm assuming if you remove pax-logging then JAX-WS webservice functionality may be affected.  If you're not using JAX-WS, it might be the case that removing pax-logging will have no negative affect.  For the time being, this is how I've chosen to proceed since I don't use JAX-WS.

**Comment by ke5stl** [comment]
hmmmm.  I would like to hear more about this JAX-WS logging dependency thing, because I am using exposing some stateless session beans as web services too, (which I expect is relying on JAX-WS) in my deployed app, and they seem to be working just fine without pax-logging.

perhaps there's some specific scenario that will be a problem while not for others?

**Comment by phillipross** [comment]
This might be a helpful reference:  Issue https://github.com/payara/Payara/issues/493

**Comment by ke5stl** [comment]
hmmm.  my current approach to this is just to remove the pax-logging module entirely, and provide all of the needed log4j2 jars and adapters (jul, slf4j, commons-logging) in the domain/lib/ext folder.  will this cause other problems?

**Comment by phillipross** [comment]
It looks like PR #672 is already in the process of fixing this issue properly.
I would suspect tossing logging framework jars into domain/lib/ext folder may create some other problems elsewhere, but if it works for you for the short term, that sounds fine.  I would just hesitate before logging any additional issues with extra problems created by the placement of these jars in those folders :smile:

**Comment by smillidge** [comment]
Yeah we are in the process of removing the dependencies by directly patching metro to remove the apache commons logging dependency which brought in pax-logging.

**Comment by phillipross** [comment]
:+1:

**Comment by ke5stl** [comment]
sounds good!  I am looking forward to this being fixed well.  the issue I was really experiencing was that the pax-logging had embedded copies of log4j code that was not compatible with the latest version/release of log4j that my deployables are using.

if use of anything other than jul can be avoided, that's probably going to be best.

**Comment by smillidge** [comment]
Please try the latest Jenkins build http://www.payara.fish/upstream_builds we think this is fixed in there.

**Comment by phillipross** [comment]
I've tested the latest pre-release from jenkins and it does indeed fix the issue.  All of my ear based apps that have slf4j+logback in the ear's lib folde are deploying normally again.

**Comment by smillidge** [comment]
@phillipross good to hear thanks for the update.

**Comment by ke5stl** [comment]
the build mentioned earlier seems to fix my logging issues too.  it would be good to have a patch release cooked up that just includes the changes for this.  I see the trunk is already moving on with other changes.  hard to know if everything from trunk is really trustworthy yet.  I have to admit though, the change to hazelcast 3.6 is good to see.

**Comment by smillidge** [comment]
I think for this bug we'll cook a patched release.

**Comment by phillipross** [comment]
@smillidge any ETA on the patched release?

**Comment by smillidge** [comment]
End of the week. It's currently in testing.

**Comment by smillidge** [comment]
Resolved in 4.1.1.161.1
